Sources/MacOSPlatform/MacOS.swift

```diff
@@ -0,0 +1,184 @@
+        guard let data = try? Data(contentsOf: infoPlist) else {
+            throw Error(message: "could not open \(infoPlist)")
+        }
+
+        guard let pkgInfo = try? decoder.decode(SwiftPkgInfo.self, from: data) else {
+            throw Error(message: "could not decode plist at \(infoPlist)")
+        }
+
+        try FileManager.default.removeItem(at: toolchainDir)
+
+        let homedir = ProcessInfo.processInfo.environment["HOME"]!
+        try runProgram("pkgutil", "--volume", homedir, "--forget", pkgInfo.CFBundleIdentifier)
+    }
+
+    public func use(_ toolchain: ToolchainVersion, currentToolchain: ToolchainVersion?) throws -> Bool {
+        let toolchainBinURL = self.swiftlyToolchainsDir
+            .appendingPathComponent(toolchain.identifier + ".xctoolchain", isDirectory: true)
+            .appendingPathComponent("usr", isDirectory: true)
+            .appendingPathComponent("bin", isDirectory: true)
+
+        // Delete existing symlinks from previously in-use toolchain.
+        if let currentToolchain {
+            try self.unUse(currentToolchain: currentToolchain)
+        }
+
+        // Ensure swiftly doesn't overwrite any existing executables without getting confirmation first.
+        let swiftlyBinDirContents = try FileManager.default.contentsOfDirectory(atPath: self.swiftlyBinDir.path)
+        let toolchainBinDirContents = try FileManager.default.contentsOfDirectory(atPath: toolchainBinURL.path)
+        let willBeOverwritten = Set(toolchainBinDirContents).intersection(swiftlyBinDirContents)
+        if !willBeOverwritten.isEmpty {
+            SwiftlyCore.print("The following existing executables will be overwritten:")
+
+            for executable in willBeOverwritten {
+                SwiftlyCore.print("  \(self.swiftlyBinDir.appendingPathComponent(executable).path)")
+            }
+
+            let proceed = SwiftlyCore.readLine(prompt: "Proceed? (y/n)") ?? "n"
+
+            guard proceed == "y" else {
+                SwiftlyCore.print("Aborting use")
+                return false
+            }
+        }
+
+        for executable in toolchainBinDirContents {
+            let linkURL = self.swiftlyBinDir.appendingPathComponent(executable)
+            let executableURL = toolchainBinURL.appendingPathComponent(executable)
+
+            // Deletion confirmed with user above.
+            try linkURL.deleteIfExists()
+
+            try FileManager.default.createSymbolicLink(
+                atPath: linkURL.path,
+                withDestinationPath: executableURL.path
+            )
+        }
+
+        SwiftlyCore.print("""
+            NOTE: On macOS it is possible that the shell will pick up the system Swift on the path
+            instead of the one that swiftly has installed for you. You can run the 'hash -r'
+            command to update the shell with the latest PATHs.
+
+                hash -r
+
+            """
+        )
```

Comment:
+1 to just doing it ourselves

Reply:
Hash is a shell builtin, so we don't really have the ability at this level to run it on behalf of the user. I'm open to better suggestions on where to put this message at this stage. Coming down the pipe with the swiftly self-installation this message gets folded into the init sub-command that only gets shown at install/upgrade time.

Reply:
This method gets called every time a different toolchain is used, so that includes even when just switching between already installed toolchains. In light of that, I'd say it'd be best to move it to `Install.swift`, and only print it after installing + using the first toolchain, only if there is already an existing Swift installation on the PATH. Since this is more of a bash thing, I don't think it is specific to macOS, so we can also drop that part of the message.

Reply:
An issue with zsh on macOS with the symlinks is that any change to the symlinks might invalidate zsh's internal path cache and it might start picking up the system's /usr/bin/swift instead of the symlink. I think that the message might be relevant any time a toolchain is used and the symlinks change. I hope that this becomes better with the proxies since the symlinks are stable until a swiftly init/self-update that can potentially change the symlinks, which is much less frequent.

Meanwhile, I think the message is annoying, but relevant at this point in time as someone might hit this in its current state.

Reply:
Have we confirmed that this happens even when just switching from one toolchain to another? It would be pretty surprising behavior if invalidating the cache caused an item later in the PATH search tree to be picked up, since while the destination of the symlink changes, the link itself will still be in the same place. Also, do we know what the macOS specific part of this is? Since it’s zsh related, I would have imagined wed see similar behavior on Linux.

Regardless, I really do think switching to a toolchain is far too common of an operation to be printing this message each time, and it runs the risk of obscuring more important output. I think if we want to be safe, printing it the first time a user installs a toolchain and including it in the ‘—help’ output is sufficient.

Reply:
I think what makes macOS so special in this respect is that its default shell is zsh, and many/most installs will have a conflicting swift binary on the user's PATH, with a dialog that can be really confusing for people using swiftly. It's this combination that I feel warrants to have a relatively loud message in this state of this commit for the sake of usability.

Reply:
If we plan to remove this message once we switch to using proxies, then the discussion is sort of moot anyways, so we can just leave it in for now. If we don't plan on removing it, then I'd like to discuss this again at that time, since I feel pretty strongly that this message is too noisy to be printed outside of installation of swiftly and/or installation of the first swiftly-managed toolchain (we already do the former actually, see [here](https://github.com/swiftlang/swiftly/blob/364b02bb2904cff62a604bfa4a0ef0d1d9164cd5/install/swiftly-install.sh#L672)).